### PR TITLE
saves hash while saving data schema

### LINKED ISSUE #164: [Data] Add the ability to create a unique hash for a dataset
**Is your feature request related to a problem? Please describe.**
To understand if data has changed in between the training of two models, we'll need some ability to take a snapshot of the data. This is of course really difficult.

**Describe the solution you'd like**
Add the ability to create a computed hash for a dataset and store that in associated to a run.

### PR DESCRIPTION
**Pull request by AnoojNair**
## Resolves
Fixes #164 

## Description
Saves Datahash while logging data schema

### CODE REVIEW

**Comment by aslotte** [comment]
This is great @AnoojNair! Could you add one more integration test with another dataset (could be the produce review data set as well with some changes) to ensure that the computed hash when the data is updated is different?

**Comment by AnoojNair** [comment]
@aslotte , done .. Had to change the getData to get all data for that run instead of just first. Is that ok?

**Comment by aslotte** [comment]
@AnoojNair you can only have one `Data` instance per run. My thought was that you can add another integration test which did something like this

```
public void Test
{
 var runId1 = await sut.LifeCycle.CreateRunAsync("test");
 var runId2 = await **sut.LifeCycle.CreateRunAsync("test");

 //then load the the different data sources and log one for run1 and another for run2, then compare that their hashes are different.
}
```

**Comment by AnoojNair** [comment]
@aslotte ,that makes total sense , sorry my bad. pushed the changes.

File: src/MLOps.NET/Catalogs/DataCatalog.cs [hunk 1]
```diff
@@ -2,6 +2,7 @@
 using MLOps.NET.Entities.Impl;
 using MLOps.NET.Storage;
 using System;
+using System.Collections.Generic;
```
Comment (aslotte) [inline src/MLOps.NET/Catalogs/DataCatalog.cs:5]: I think is not needed anymore

File: src/MLOps.NET/Entities/Impl/Data.cs [hunk 2]
```diff
@@ -31,5 +31,10 @@ public Data(Guid runId)
         /// DataSchema
         /// </summary>
         public DataSchema DataSchema { get; set; }
+
+        /// <summary>
+        ///
```
Comment (aslotte) [inline src/MLOps.NET/Entities/Impl/Data.cs:36]:
```suggestion
        /// DataHash
```

File: src/MLOps.NET/Storage/Interfaces/IDataRepository.cs [hunk 3]
```diff
@@ -1,6 +1,7 @@
 ﻿using Microsoft.ML;
 using MLOps.NET.Entities.Impl;
 using System;
+using System.Collections.Generic;
```
Comment (aslotte) [inline src/MLOps.NET/Storage/Interfaces/IDataRepository.cs:4]: This is probably not needed anymore

File: src/MLOps.NET/Storage/MetaDataStore/DataRepository.cs [hunk 4]
```diff
@@ -46,6 +52,22 @@ public async Task LogDataAsync(Guid runId, IDataView dataView)
...
+        private string GetStringHashFromDataView(IDataView dataView)
+        {
+            string hash = "";
+            using (SHA256 sha256Hash = SHA256.Create())
+            {
+                using (var stream = new MemoryStream())
+                {
+                    new MLContext().Data.SaveAsBinary(dataView, stream);
+                    var bytes = sha256Hash.ComputeHash(stream);
+                    hash = Convert.ToBase64String(bytes);
```
Comment (aslotte) [inline src/MLOps.NET/Storage/MetaDataStore/DataRepository.cs:64]: No need to initialize a separate variable here, we can just do `return Convert.ToBase64String(bytes)`

File: test/MLOps.NET.IntegrationTests/DataCatalogTests.cs [hunk 5]
```diff
@@ -36,11 +36,52 @@ public async Task LogDataAsync_GivenValidDataView_ShouldLogData()
...
+
+            var updatedData = LoadUpdatedData();
+
+            //Act
+            await sut.Data.LogDataAsync(previousRunId, data);
+            await sut.Data.LogDataAsync(currentRunId, updatedData);
+            var previousRunData = sut.Data.GetData(previousRunId);
+            var currentRunData = sut.Data.GetData(currentRunId);
+            //Assert
+
```
Comment (aslotte) [inline test/MLOps.NET.IntegrationTests/DataCatalogTests.cs:70]:
```suggestion
```

File: test/MLOps.NET.IntegrationTests/DataCatalogTests.cs [hunk 6]
```diff
@@ -36,11 +36,51 @@ public async Task LogDataAsync_GivenValidDataView_ShouldLogData()
...
```
Comment (aslotte) [inline test/MLOps.NET.IntegrationTests/DataCatalogTests.cs:70]:
```suggestion
```

File: test/MLOps.NET.IntegrationTests/DataCatalogTests.cs [hunk 7]
```diff
@@ -36,11 +36,50 @@ public async Task LogDataAsync_GivenValidDataView_ShouldLogData()
...
+            var data = LoadData();
+
+            var updatedData = LoadUpdatedData();
+
+            //Act
+            await sut.Data.LogDataAsync(previousRunId, data);
+            await sut.Data.LogDataAsync(currentRunId, updatedData);
+            var previousRunData = sut.Data.GetData(previousRunId);
+            var currentRunData = sut.Data.GetData(currentRunId);
+            //Assert
```
Comment (aslotte) [inline test/MLOps.NET.IntegrationTests/DataCatalogTests.cs:69]:
```suggestion

            //Assert
```

File: test/MLOps.NET.IntegrationTests/DataCatalogTests.cs [hunk 8]
```diff
@@ -36,11 +36,51 @@ public async Task LogDataAsync_GivenValidDataView_ShouldLogData()
...
+            savedData.DataHash.Should().NotBeNullOrEmpty();
+        }
+
+        [TestMethod]
+        public async Task LogDataAsync_ShouldGenerateDifferentHashWhenDataChanges()
+        {
+            var previousRunId = await sut.LifeCycle.CreateRunAsync("previous run");
+            var currentRunId = await sut.LifeCycle.CreateRunAsync("current run");
+            var data = LoadData();
+
```
Comment (aslotte) [inline test/MLOps.NET.IntegrationTests/DataCatalogTests.cs:61]:
```suggestion
```

File: test/MLOps.NET.IntegrationTests/DataCatalogTests.cs [hunk 9]
```diff
@@ -36,11 +36,51 @@ public async Task LogDataAsync_GivenValidDataView_ShouldLogData()
...
+            savedData.DataHash.Should().NotBeNullOrEmpty();
+        }
+
+        [TestMethod]
+        public async Task LogDataAsync_ShouldGenerateDifferentHashWhenDataChanges()
+        {
+            var previousRunId = await sut.LifeCycle.CreateRunAsync("previous run");
+            var currentRunId = await sut.LifeCycle.CreateRunAsync("current run");
+            var data = LoadData();
```
Comment (aslotte) [inline test/MLOps.NET.IntegrationTests/DataCatalogTests.cs:60]:
```suggestion

            var data = LoadData();
```

File: test/MLOps.NET.Tests.Common/MLOps.NET.Tests.Common.csproj [hunk 10]
```diff
@@ -5,6 +5,10 @@
     <IsPackable>false</IsPackable>
   </PropertyGroup>
 
+  <ItemGroup>
+    <None Remove="Data\product_reviews_updated.csv~RFc72d885.TMP" />
```
Comment (aslotte) [inline test/MLOps.NET.Tests.Common/MLOps.NET.Tests.Common.csproj:9]: I don't think we need this

**Comment by AnoojNair** [comment]
Thank you @aslotte , pushed the changes.

**Pull request merged**